components/x-audio/src/redux/__tests__/middleware/notifier.test.js

```diff
@@ -0,0 +1,119 @@
+import middleware, { NotifiersProxy } from '../../middleware/notifier';
+import { actions, initialState } from '../../player-logic';
+
+
+const create = (currentState = {}) => {
+	const state = { ...initialState, ...currentState };
+	const store = {
+		getState: () => state,
+		dispatch: jest.fn()
+	}
+	const next = jest.fn();
+
+	const notifier = {
+		play: jest.fn(),
+		pause: jest.fn(),
+		ended: jest.fn(),
+		tracking: jest.fn()
+	}
+
+	const middlewareWithNext = middleware(notifier)(store)(next);
+	const invoke = action => middlewareWithNext(action)
+	return { store, next, invoke, notifier };
+}
+
+describe('Notifier middleware', () => {
+
+	test('just calls next for unknown actions', () => {
+		const { invoke, next } = create();
+		const action = { type: 'OTHER_PLAYER_ACTION' }
+		invoke(action);
+		expect(next).toBeCalledWith(action);
+	});
+
+	[
+		[ actions.play(), 'play' ],
+		[ actions.pause(), 'pause' ],
+		[ actions.ended(), 'ended' ],
+	].forEach(([ action, notifyFn ]) => {
+		test(`${action.type} action triggers ${notifyFn} notification`, () => {
+			const { invoke, notifier, next } = create();
+			invoke(action);
+			expect(notifier[notifyFn]).toHaveBeenCalled();
+			expect(next).toBeCalledWith(action);
+		});
+	});
+
+	test('external PLAY action does not trigger play notification', () => {
+		const { invoke, notifier, next } = create({ isPlayInternal: false });
+		const action = actions.play();
+		invoke(action);
+		expect(notifier.play).not.toHaveBeenCalled();
+		expect(next).toBeCalledWith(action);
+	});
+	test('external PAUSE action does not trigger pause notification', () => {
+		const { invoke, notifier, next } = create({ isPauseInternal: false });
+		const action = actions.pause();
+		invoke(action);
+		expect(notifier.pause).not.toHaveBeenCalled();
+		expect(next).toBeCalledWith(action);
+	});
+
+	describe('tracking', () => {
+		const state = { loading: true };
+
+		[
+			[ actions.requestPlay(), 'play', ['play-click', state ] ],
+			[ actions.requestPause(), 'pause', ['pause-click', state ]],
+			[ actions.willClose(), 'close', ['close-click', state ] ],
+		].forEach(([ action, notifyFn, payload ]) => {
```

Comment:
Appreciate that this makes the tests DRY and the console output will be neat but I wonder if its worth just breaking these out...

Could potentially have a helper function like `fireAction()`? Then expect the relevant notifier to have called?

Reply:
In fact `fireAction` would be pointless. But still wonder if its worth just duplicating them for the sake of simplicity in the tests

Reply:
Yeah know what you mean. Main intention of doing it was to make sure `next` is being called each time, cos it's not immediately clear from the code you need. Thought about using `before` and `after` hooks but got a bit messy. If we split it out, then future tests will just need to remember to check next has been called...which might be ok??!

Reply:
Yeah, I think on reading it again it's fine as is. It was just the whole tests as documentation thing, which this does do!

Reply:
Split it out 👍